tlb: kernel after write tlb need flush

zgyxyd666 opened the pull request:
Without a flush, stale TLB entries may still reference previous mappings, causing
incorrect address translations during early kernel initialization.

## Change Details
Add `axhal::asm::flush_tlb(None);` immediately after `write_kernel_page_table()`.

AsakuraMizu committed:
fix:  customizable *_feat_prefix

LIN-Matrix committed:
refactor: axnet

Co-authored-by: mivik <mivikq@gmail.com>

Mivik committed:
feat(wq): add notify_all_if & support FnMut in wait_until

Mivik committed:
fix(task): task entry is not dropped

Co-authored-by: mivik <mivikq@gmail.com>

Mivik committed:
fix(mm): address overflow

Mivik committed:
feat(mm): new protect strategy

Mivik committed:
refactor: update axerrno

Mivik committed:
feat: use new scheduler, fix(?) preempt slowdown

Mivik committed:
feat: add experimental future support

replace axsync::Mutex implementation & some other WaitQueue usage
with event_listener::Event

AsakuraMizu committed:
refactor: axmm

- better backend impl
- copy feature
- try_clone
- populate_area
- protect

Thanks to: oscomp#4, oscomp#5, oscomp#11, oscomp#14, oscomp#15, oscomp#16

Co-authored-by: mivik <mivikq@gmail.com>
Co-authored-by: Youjie Zheng <Azure_stars@126.com>

AsakuraMizu committed:
feat: patch axcpu

AsakuraMizu committed:
feat(task): expose more information

Co-authored-by: Youjie Zheng <Azure_stars@126.com>

AsakuraMizu committed:
chore: cleanup workspace

Mivik committed:
refactor(task): move AxWaker

AsakuraMizu committed:
feat(mm): shared backend

Co-authored-by: mivik <mivikq@gmail.com>

Mivik committed:
feat(net): utilize smoltcp async feature

Mivik committed:
refactor: axfs-ng

Co-authored-by: Asakura Mizu <asakuramizu111@gmail.com>

AsakuraMizu committed:
fix allocator & override memory size

Co-authored-by: mivik <mivikq@gmail.com>

AsakuraMizu committed:
refactor: task-ext and scope-local

AsakuraMizu committed:
feat: backtrace

Co-authored-by: mivik <mivikq@gmail.com>

Mivik committed:
refactor axnet

Mivik committed:
refactor(net): unify the use of InterfaceWrapper

Mivik committed:
feat(net): basic eth0 support

AsakuraMizu committed:
fix(mm): can_access_range

AsakuraMizu committed:
perf(alloc): mem usage stats

Mivik committed:
feat(net): upgrade smoltcp, add routing & reuseaddr support

Mivik committed:
fix(fs/ext4): creating existing entry should raise an error

Mivik committed:
feat(net): use axio::Buf

Mivik committed:
chore: sync lwext4_rust

Mivik committed:
chore: sync axfs-ng-vfs

Mivik committed:
feat: add pre_trap, task interruption & fix socket

AsakuraMizu committed:
chore(mm): simplify api

AsakuraMizu committed:
chore(task): use AtomicWaker

AsakuraMizu committed:
feat: improve trap backtrace

Mivik committed:
feat(mm): add huge page & COW support

Mivik committed:
refactor: replace BTreeMap with hashbrown

Mivik committed:
chore: move lazy_static to workspace dependencies

AsakuraMizu committed:
fix(task): don't resched on wake

AsakuraMizu committed:
fix(mm): compile error on LA

Mivik committed:
feat(fs): add page cache

AsakuraMizu committed:
chore(task): expose api

Mivik committed:
feat: add page cache evict listener, refactor mm backend, refactor tmpfs

AsakuraMizu committed:
fix(mm): compile error on la

AsakuraMizu committed:
feat(task): asyncify exit

Mivik committed:
fix(net): neighbor cache expiry

AsakuraMizu committed:
feat(task): timer callback

Mivik committed:
feat(fs): bind page cache to DirEntry

Mivik committed:
fix: page cache does not recycle

Mivik committed:
fix(mm): page cache & file backend ownership problem

Mivik committed:
fix(mm): COW backend should also use FileBackend

Mivik committed:
fix: in memory file evict cache

AsakuraMizu committed:
refactor(hal): redesign user mode

update axcpu

AsakuraMizu committed:
fix(task): drop task entry properly

related: 3958dd1a2d9e91ebaa613bcea104331383a56d2a

AsakuraMizu committed:
feat: open_loc

also format code

Mivik committed:
chore: sync axfs-ng-vfs

AsakuraMizu committed:
fix(fs): append & seek

Mivik committed:
fix(fs): sync lwext4, don't write discarded dirty pages

Mivik committed:
style: fix clippy

Mivik committed:
feat(axhal): remove unused link section

Mivik committed:
feat: new interruption mechanism, add try_block_on

Mivik committed:
feat(fs): add O_PATH, fix O_APPEND

Mivik committed:
refactor(mm): export strcuture

AsakuraMizu committed:
chore: improve qemu debug

Mivik committed:
feat: add unix address support

Mivik committed:
refactor(net): simplify tcp structure

Mivik committed:
feat(net): add basic UDS support

Mivik committed:
feat(net): add basic cmsg support

Mivik committed:
fix: remove axio patch

Mivik committed:
chore: update Cargo.lock

Mivik committed:
fix(net): UDS close behavior

Mivik committed:
fix(task): AxWaker should not strong reference AxTask, also fix gc task

Mivik committed:
refactor(net): change UDS to static dispatch

Mivik committed:
refactor: BREAKING poll machanism

also includes:

- axnet loopback polling fix (and possibly enhancement)
- remove strong ref to `AxTask` in `TaskWakeupEvent`

AsakuraMizu committed:
feat: generally improve robustness

Mivik committed:
fix(mm): use weak ref to `AddrSpace` in `FileBackend`

Mivik committed:
fix(mm): region error type

Mivik committed:
fix(mm): mmap error

AsakuraMizu committed:
refactor(fs): remove lock_api trait

AsakuraMizu committed:
chore: update lockfile

Mivik committed:
feat(net/unix): add SNDBUF

Mivik committed:
feat: add display, sync arceos

Mivik committed:
chore: sync lwext4-rust

Mivik committed:
refactor(net): unify socket polling

Mivik committed:
feat(net/unix): add ucred

Mivik committed:
feat(net): add RDHUP & fix notify

Mivik committed:
fix(net/unix): polling

AsakuraMizu committed:
chore: cleanup debug sections

Mivik committed:
fix(net): non-blocking unix stream & listen

AsakuraMizu committed:
feat(driver): sdmmc with gpt root

AsakuraMizu committed:
fix(build): objcopy binary name

Mivik committed:
feat(mm): add read_only option to CowBackend

AsakuraMizu committed:
feat(alloc): tracking switch

Mivik committed:
feat(mm): adapt to new page table API

Mivik committed:
feat: add icount

Mivik committed:
fix(build): la axmm

AsakuraMizu committed:
chore(mm): cleanup code

AsakuraMizu committed:
fix(task): reimplement timer list

Mivik committed:
feat: support node flags

AsakuraMizu committed:
fix(alloc): memory tracker crash

AsakuraMizu committed:
fix(mm): user read-only memory not recycled

`read_only` introduced in https://github.com/Starry-Mix-THU/arceos/commit/b00a71cf7dc9832d7b73669caee013afc9ef2709 causes memory leak. Read-only memory need to be ref-counted and recycled as well!
Also do some code cleanup.

AsakuraMizu committed:
replace elf helper with bash script

AsakuraMizu committed:
fix(build): run dwarf script only if backtrace enabled

Mivik committed:
feat: add input devices

Mivik committed:
fix(fs): multiple user_data

AsakuraMizu committed:
chore(driver): unify behav of dyn and static

Mivik committed:
feat: add fs flush, sync vfs & lwext4

Mivik committed:
feat: add PLIC support (net)

Mivik committed:
feat(driver): make IRQ optional

AsakuraMizu committed:
chore: remove log

Mivik committed:
fix: gc_task preempt

Mivik committed:
feat(axhal): re-export console::enable_rx_interrupt

AsakuraMizu committed:
build: pin git deps

Mivik committed:
chore: sync lwext4_rust

Mivik committed:
chore: sync axio, lower IRQ log level

Mivik committed:
chore: sync axcpu

Mivik committed:
feat: add spawn task name, sync axcpu

Mivik committed:
refactor: remove WaitQueue

Mivik committed:
chore: sync axcpu

Mivik committed:
feat(fs): add times & larger ext4 bcache

Mivik committed:
fix(task): timeout logic

AsakuraMizu committed:
sync axplat

Mivik committed:
feat: add fancy banner

AsakuraMizu committed:
chore: axbacktrace fp range

Mivik committed:
feat(fs): make File::new public

AsakuraMizu committed:
build(la): disable ual

AsakuraMizu committed:
chore: log

Mivik committed:
feat(hal): add exception table

AsakuraMizu committed:
chore: custom QEMU_ARGS

Mivik committed:
feat: adapt to new axio design

AsakuraMizu committed:
feat(driver): fixed address for ramdisk

Mivik committed:
fix(net): missing poll

AsakuraMizu committed:
fix: re-export axcpu stuffs

AsakuraMizu committed:
feat: ramdisk config

AsakuraMizu committed:
feat(driver,shit): ahci-gpt

AsakuraMizu committed:
feat: lto build option

AsakuraMizu committed:
Revert "feat: add fancy banner"

This reverts commit 56e8a6020ad8b4f1cb21a5496d4b6e3f723ca37e.

AsakuraMizu committed:
bump dependencies

AsakuraMizu committed:
Merge remote-tracking branch 'upstream/main'

AsakuraMizu committed:
feat(la64-qemu): boot from binary

Mivik committed:
feat: adapt to new axerrno

AsakuraMizu committed:
chore: update deps

AsakuraMizu committed:
fix(hal): remove aarch64 sp_el0 hack

AsakuraMizu committed:
fix(driver): update deps

AsakuraMizu committed:
update deps

AsakuraMizu committed:
update axplat

AsakuraMizu committed:
refactor(task): future

AsakuraMizu committed:
refactor: axpoll

eternalcomet committed:
chore: update deps

AsakuraMizu committed:
feat(fs): convenience methods for I/O

guoweikang committed:
feat: vsock (#3)

* axdriver: add vsock driver support

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* axnet: add AF_VSOCK socket support

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* axruntime: add vsock feature

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* arceos: add virtio-vsock qemu config

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* vsock: add vsock feature for axnet

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* vsock: Unified terminology using VSOCK

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* vsock: Log degradation

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* vsock:lazy_static: remove lazy_static use

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

* chore: cleanup code

---------

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>
Co-authored-by: Weikang Guo <guoweikang@kylinos.cn>
Co-authored-by: Asakura Mizu <asakuramizu111@gmail.com>

AsakuraMizu committed:
chore: cleanup code for plat support; removing 2k1000la stuffs

guoweikang committed:
tlb: kernel after write tlb need flush

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>

equation314 commented on modules/axmm/src/lib.rs:
```diff
@@ -106,6 +106,8 @@ pub fn init_memory_management() {
     KERNEL_ASPACE.init_once(SpinNoIrq::new(kernel_aspace));
     unsafe { axhal::asm::write_kernel_page_table(kernel_page_table_root()) };
+    // flush all tlb
+    axhal::asm::flush_tlb(None);
```
Also add it in `init_memory_management_secondary`. And the flush is not required on x86_64.

equation314 reviewed (commented):
Yes, TLB flushing was missed after https://github.com/arceos-org/arceos/pull/257. Thanks!

AsakuraMizu committed:
chore: remove patch for axbacktrace and axpoll

AsakuraMizu committed:
chore: do not pin the version of log crate

AsakuraMizu committed:
chore: update axerrno and other deps

AsakuraMizu committed:
chore: update page_table_multiarch, use type conversion

AsakuraMizu committed:
chore: update allocator and memory_set

alvinsun039 committed:
fix(axsync): deadlock from spurious weak CAS failure

Skip block_on when the lock is actually unlocked to avoid waiting
forever with no one to wake us. This handles spurious failures
from compare_exchange_weak on LL/SC architectures.

Signed-off-by: Ke Sun <sunke@kylinos.cn>
Co-authored-by: Ke Sun <sunke@kylinos.cn>

AsakuraMizu committed:
fix(axsync): rewrite mutex

Qzincs committed:
fix(net/vsock): allow reading remaining data after connection closed (#6)

Signed-off-by: Weikang Guo <guoweikang@kylinos.cn>
Co-authored-by: Weikang Guo <guoweikang@kylinos.cn>

Commit:
mm(axmm): add secondary TLB flush (skip x86_64)

Commit:
Merge upstream main to resolve conflicts
